### ISSUE TITLE
[RFC]: Extended online quantization roadmap

### ISSUE THREAD

**Issue by fxmarty-amd** [vllm-project/vllm#52167]
### Motivation.

Continuation of https://github.com/vllm-project/vllm/issues/32029.

Online quantization API is now relatively stable at https://github.com/vllm-project/vllm/tree/main/vllm/model_executor/layers/quantization/online and documented at https://docs.vllm.ai/en/stable/features/quantization/online/.

This RFC summarizes and motivates proposed features that would be helpful to make online quantization to be more flexible and more widely compatible with models, including pre-quantized ones.

### 1. Support activation quant key override

Implemented in https://github.com/vllm-project/vllm/pull/51808.

This is necessary to dispatch on specific linear/moe backends that can be picked only through the `activation_quant_key` (see e.g. https://github.com/vllm-project/vllm/blob/c4e969294ecab9ffefb995b734303f13f62b723f/vllm/model_executor/layers/fused_moe/oracle/mxfp4.py#L441-L445)

### 2. Targeted online quantization configuration based on user patterns

Implemented in https://github.com/vllm-project/vllm/pull/51285, RFC in https://github.com/vllm-project/vllm/issues/50281.

The goal is simply to extend the UI to support e.g. `--quantization-config.targets '{"re:.*o_proj.*":"mxfp8"}'`, e.g. to target a single layer similar to e.g. https://github.com/vllm-project/vllm/blob/dc9f845ddc54c1df38fdbce5afe03f9fd15813bd/vllm/model_executor/layers/attention/mla_attention.py#L935-L948 (that is currently hard-coded and may be causing accuracy issues if done silently https://github.com/vllm-project/vllm/pull/48051), or e.g. `--quantization-config.targets '{"re:.*shared_expert.*":"fp8_per_block"}'` to enable fused shared expert.

### 3. Support online quantization with partially pre-quantized checkpoints

Implemented in https://github.com/vllm-project/vllm/pull/51392.

This is necessary to combine online quantization with already partially quantized checkpoints, which is notoriously the case recently. This is e.g. useful to test the accuracy/speed tradeoff of quantizing certain additional layers, or useful to e.g. release a single model compatible with fused shared expert / unfused shared expert.

This has e.g. been requested in separate PRs:
- https://github.com/vllm-project/vllm/pull/48983
- https://github.com/vllm-project/vllm/pull/49313

### 4. Requantize already-quantized layers to different precision (different weight quant key)

Not implemented yet, closest prototype is https://github.com/vllm-project/vllm/pull/48427.

This is similar to https://github.com/sgl-project/sglang/pull/28291 & https://github.com/sgl-project/sglang/pull/29328 on sglang side.

This idea here would be to allow loading through a certain quantization method, dequantize and override the quantization method on request.

This has e.g. been requested in:
- https://github.com/vllm-project/vllm/pull/48427
- https://github.com/vllm-project/vllm/pull/51274

### 5. Adopt `online/*` API for Humming online quantization, torchao online quantization

Currently, Humming (https://github.com/vllm-project/vllm/pull/34556) and torchao (https://github.com/vllm-project/vllm/pull/32133, https://github.com/vllm-project/vllm/pull/23014) have their own online quantization logic:

https://github.com/vllm-project/vllm/blob/93eea4f665a0a6a188fd52b37f4a18bc37315f62/vllm/model_executor/layers/quantization/humming.py#L268-L277

https://github.com/vllm-project/vllm/blob/340b7e4909f14ff33943ab6ac7955191e1b8b4ff/vllm/model_executor/layers/quantization/torchao.py#L383-L389

In contrary proposed online quantization for AMD/Quark in vLLM that intended to plug similarly in existing quant method was rejected: https://github.com/vllm-project/vllm/issues/31028 - so we ended up [having a plugin](https://github.com/amd/Quark/tree/release/0.12/examples/online_quantization) (and eventually adopting upstream `online/*` API now).

These torchao/humming logics could be moved to use the online quantization frontend.

cc @jerryzh168 @kylesayrs @jinzhen-lin 

### Feedback Period.

One month

### CC List.

@BowenBao @mgoin @kylesayrs @vkuzo @maeehart @tanpinsiang @fanxingran @access2rohit

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by 200lz** [comment]
One thing that may be worth making explicit as requantization lands is the resolved transformation chain, not just the final quant method. For example: source tensor/quantization → dequantization method → requested online quantization → resolved backend/representation. That would make it much easier to distinguish “checkpoint as published” from a runtime-derived numerical representation when debugging accuracy changes.

**Comment by fxmarty-amd** [comment]
> One thing that may be worth making explicit as requantization lands is the resolved transformation chain, not just the final quant method. For example: source tensor/quantization → dequantization method → requested online quantization → resolved backend/representation. That would make it much easier to distinguish “checkpoint as published” from a runtime-derived numerical representation when debugging accuracy changes.

For sure, especially given that at the moment there does not seem to be 1:1 numerical agreement for some online schemes, see https://github.com/vllm-project/vllm/pull/46400. It was discussed a while ago in a SIG quantization call (does not seem too bad, but e.g. issue of positive/negative zeros, similar to: https://github.com/vllm-project/vllm/blob/1f7427bc0ad31fdad1b0933c77ea5acae15da2e0/tests/quantization/test_online_mxfp4.py#L45 )

https://github.com/vllm-project/vllm/pull/51285 and https://github.com/vllm-project/vllm/pull/51392 already include some logging like `INFO 08-06 16:35:33 [base_loader.py:106] Quantized 48 layers of types: block_sparse_moe.experts: 24 (from targets: re:.*block_sparse_moe\.experts, fp8_per_tensor); self_attn.o_proj: 24 (from targets: re:.*self_attn\.o_proj, fp8_per_block)` that should already make it a bit more transparent.

But there is definitely a question of how to have `--moe-backend`, `--linear-backend` compatible with mixed linear/moe precisions for different layers. I think it is an existing issue for offline quantization methods as well, though (similar issue for `--attention-backend` in connection with https://github.com/vllm-project/vllm/pull/47896#discussion_r3588688397)

> source tensor/quantization → dequantization method 

I think the idea in https://github.com/vllm-project/vllm/pull/48427 to have the dequantization method part of the original quant method (and `NotImplementedError` otherwise) is sound.

------

Note: one issue about `shared_expert` + fusion. Current implementation, e.g. https://github.com/vllm-project/vllm/blob/1f7427bc0ad31fdad1b0933c77ea5acae15da2e0/vllm/model_executor/models/qwen3_next.py#L168-L180

is not very well compatible with online quantization of shared expert.

**Comment by 200lz** [comment]
Thanks, this is very helpful. The existing summary log covers targeting intent; the remaining gap seems to be the resolved path per module: checkpoint quant method → dequantizer → requested online scheme → selected backend/kernel and representation, including fallbacks.

The ±0 example is also a good reason to separate bitwise representation from dequantized numerical equivalence. I can prototype a compact machine-readable record and validate it on one offline-vs-online quantization path if that would be useful.

**Comment by maeehart** [comment]
Item 4 is the right place for MXFP4-to-groupwise-INT4 at load. Draft #51274 is a Kimi-K3 prototype of that (`--quantization-config.moe.weight int4_per_group_32`). Qwen3.8 needs the same thing, and more MXFP4 MoE models will too.

I think that we should not keep publishing extra Hugging Face checkpoints per model. That becomes a bottleneck. One opt-in at-load conversion onto an INT4 expert kernel is enough, and it belongs in this RFC rather than as a pile of model-specific quantizations.

gfx942 tracker that depends on this: https://github.com/vllm-project/vllm/issues/52803

**Comment by fxmarty-amd** [comment]
@maeehart agreed, especially when it is a something like a single layer dtype change through RTN quantization (e.g. shared_expert) for very large models.

The thing is, given the requests we have seen for this, I am thinking it'd be better to make it as much as possible quant-method / model-producer agnostic, thus `4. Requantize already-quantized layers to different precision (different weight quant key)` in this RFC.

You might be interested in seeing https://github.com/fxmarty-amd/vllm/pull/7 that is a showcase draft adapted from https://github.com/vllm-project/vllm/pull/48427 (that I like, very extendible) showing how it integrates IMO nicely with proposed https://github.com/vllm-project/vllm/pull/51392

**Comment by fxmarty-amd** [comment]
Hi there!

https://github.com/vllm-project/vllm/pull/51392 and https://github.com/vllm-project/vllm/pull/51285 have been merged, now:

## Support activation quant key override

is implemented in https://github.com/vllm-project/vllm/pull/51808, needs review.

## Requantize already-quantized layers to different precision (different weight quant key)

is implemented in https://github.com/vllm-project/vllm/pull/55684, needs review as well. To give more context about it:

There have been proposal as https://github.com/vllm-project/vllm/pull/48427 and https://github.com/vllm-project/vllm/pull/51274 to allow weight re-quantization in vLLM (changing their quant key), from e.g.:

- `MXFP8` to `FP8 PTPC`
- `MXFP4` to `INT4`.

This is something also available in sglang at https://github.com/sgl-project/sglang/pull/28291 and https://github.com/sgl-project/sglang/pull/29328, see e.g. https://rocm.blogs.amd.com/software-tools-optimization/nvfp4-to-mxfp4/README.html

My opinion is that:

- attempting to have a given quant key oracle (say `oracle/mxfp4.py`) handling weight re-quantization to an other quant key in e.g. `convert_weight_to_mxfp4_moe_kernel_format` is not a good thing to do. Instead, we should have composability with, say, an `int4`/`int_wna16` oracle in case we attempt to re-quantize to int4 as in https://github.com/vllm-project/vllm/pull/51274. Which allows to then use any INT4 backend, instead of hacking our way into an oracle that is not meant to handle dispatch for an other weight quant key.
- This is the same thing in https://github.com/vllm-project/vllm/pull/51808 (`MXFP8 -> FP8 PTPC`). Instead of hacking a new FP8 PTPC backend into MXFP8 oracle, we just use FP8 oracle to dispatch to whichever backend we'd like to use after re-quantization.
- We should maintain strict separation between oracle and quant methods (see https://github.com/vllm-project/vllm/issues/54959)

Note how lightweight the proposed implementation is in https://github.com/vllm-project/vllm/pull/55684:

```python
    def dequantize_weight(self, layer: torch.nn.Module) -> torch.Tensor:
        """Reconstruct serialized weights for online requantization."""
        if self.wkey.key is kMxfp8Static:
            return dequant_mxfp8_to_bf16(
                layer.weight.contiguous(), layer.weight_scale.contiguous()
            )
        else:
            raise NotImplementedError(
                "ModelOpt weight dequantization is only supported for MXFP8."
            )
```

that is completely optional, as the base `QuantizeMethodBase` defines:

```python
    def dequantize_weight(self, layer: nn.Module) -> torch.Tensor:
        """Materialize a serialized quantized weight for requantization."""
        raise NotImplementedError(
            f"The quantization method {type(self)} does not implement "
            "dequantize_weight. Please open an issue."
        )
```

Essentially, I don't think we should reason in a way where `BF16/FP16/FP32` dtypes have a privileged role compared to other dtypes when we think about online quantization (beyond being the interface, layer by layer at load time, between `dtypeA -> BF16 -> dtypeB`, or `dtypeA -> FP32 -> dtypeB`).

`FP8 -> MXFP4` casting in vLLM should STILL be considered online quantization. Same as `NVFP4 -> MXFP4`, or anything else.

Note: I believe this topic is strongly related to https://github.com/vllm-project/vllm/issues/37753. This proposal will work smoothly only if the oracles handling different quant key have a unified API. For example, there should not be a single oracle handling Kimi-K3 specific `MoEActivation.SITU` item.

Long term, it should make maintainability a lot easier to maintain a clear separation between:

- model producers (Quark/modelopt/compressed-tensors/mxfp4/fp8) / quant method classes handling **loading only**,
- oracles handling different weight key (that should hopefully be unified as per https://github.com/vllm-project/vllm/issues/37753),
- MOE/linear backends handling specific weight key, and dispatched to by their oracle.
- have online quantization be the interface to cast to different weight key if we want to.
